vignettes/oauth.Rmd

```diff
@@ -0,0 +1,479 @@
+In a real package you could cache the OAuth2.0 tokens on the machine, e.g. using the keyring package.
+
+```{r eval = FALSE}
+gh_repos()
+```
+
+```{r echo = FALSE, eval = TRUE}
+# Ideally we would only run this when we don't have a real
+# secret, but httr's OAuth does not work well in non-interactive
+# sessions.
+message('Waiting for authentication in browser...')
+message('Press Esc/Ctrl + C to abort')
+message('Authentication complete.')
+print(c(
+  "gaborcsardi/altlist", "gaborcsardi/argufy",
+  "gaborcsardi/async", "gaborcsardi/disposables",
+  "gaborcsardi/dotenv", "gaborcsardi/falsy",
+  "gaborcsardi/franc", "gaborcsardi/ISA",
+  "gaborcsardi/keypress", "gaborcsardi/lpSolve",
+  "gaborcsardi/macBriain", "gaborcsardi/maxygen",
+  "gaborcsardi/MISO", "gaborcsardi/msgtools",
+  "gaborcsardi/notifier", "gaborcsardi/odbc",
+  "gaborcsardi/parr", "gaborcsardi/parsedate",
+  "gaborcsardi/prompt", "gaborcsardi/r-font",
+  "gaborcsardi/r-source", "gaborcsardi/rcorpora",
+  "gaborcsardi/roxygenlabs", "gaborcsardi/sankey",
+  "gaborcsardi/secret", "gaborcsardi/spark",
+  "gaborcsardi/standalones"
+))
+```
+
+Let's write a test case now for `gh_repos()`.
+We will use `oauth2_repource_app()` to fake GitHub.
+
+```{r}
+testthat::test_that("gh_repos", {
+  testthat::skip_on_cran()
+  fake_app <- oauth2_resource_app(token_endpoint = "/access_token")
+  fake_app$get("/user/repos", function(req, res) {
+    if (!app$is_authorized(req, res)) return()
+    res$send_json(list(
+      list(full_name = "user/repo1"),
+      list(full_name = "user/repo2")
+    ), auto_unbox = TRUE)
+  })
+
+  fake_proc <- local_app_process(
+    fake_app,
+    opts = server_opts(num_threads = 3)
+  )
+
+  # register the app to our fake GH server
+  reg_url <- paste0(
+    fake_proc$url("/register"),
+    "?name=gh_repos&redirect_uri=http://localhost:1410/"
+  )
+  regdata <- httr::content(httr::GET(reg_url))
+  withr::local_envvar(
+    GH_CLIENT_ID = regdata$client_id[[1]],
+    GH_CLIENT_SECRET = regdata$client_secret[[1]],
+    FAKE_GH_API_BASE = fake_proc$url(),
+    FAKE_GH_AUTH_BASE = fake_proc$url()
+  )
+
+  ret <- suppressMessages(webfakes::oauth2_httr_login(
+    gh_repos()
+  ))
+  testthat::expect_equal(ret, c("user/repo1", "user/repo2"))
+})
+```
+
+Some important points about this test case:
+
+-   Since the test case will fail if port 1410 is taken, it is safer to skip it on CRAN.
+
+-   `fake_app` is a webfakes app, that is almost the same as `oauth2_resource_app()`, with two changes.
+
+    -   The first is the we change the end point for getting a token to `/access_token` because that is what GitHub uses.
+    -   The second is that we also add the `/user/repos` endpoint. This endpoint needs authorization, this is why it calls `app$is_authorized()`, which fails without it, and the app returns 401 Access denied.
+
+-   `fake_proc` is the app process that runs our fake GH app. Always run the fake OAuth2.0 apps with multiple threads.
```

Comment:
Can you clarify "Always run the fake OAuth2.0 apps with multiple threads."?

Reply:
It is a technical detail of the current implementation of the fake apps, so I would just omit the explanation here.